packages/runtime/src/events/handleAction.ts

```diff
@@ -91,23 +93,54 @@ export function handleAction(
             toddle: ctx.toddle,
             env: ctx.env,
           })
+          // historyMode was previously not declared explicitly, and we default
+          // to push for state changes and replace for query changes
+          let historyMode: SetURLParameterAction['historyMode'] | undefined
+          let newLocation: Location | undefined
+          // We should only match on p.type === 'param', but
+          // that would technically be a breaking change
           if (current.route?.path.some((p) => p.name === action.parameter)) {
-            return {
+            historyMode = 'push'
+            newLocation = {
               ...current,
               params: {
                 ...omitKeys(current.params, [action.parameter]),
                 [action.parameter]: value,
               },
             }
-          } else {
-            return {
+          }
+          // We should check if the query parameter exists in the route
+          // but that would technically be a breaking change
+          // else if (Object.values(current.route?.query ?? {}).some((q) => q.name === action.parameter))
+          else {
+            historyMode = 'replace'
+            newLocation = {
               ...current,
               query: {
                 ...omitKeys(current.query, [action.parameter]),
                 ...(isDefined(value) ? { [action.parameter]: value } : null),
               },
             }
           }
+          if (!historyMode) {
+            // No path/query parameter matched
+            return current
+          }
+
+          const currentUrl = getLocationUrl(current)
+          const historyUrl = getLocationUrl(newLocation)
+          if (historyUrl !== currentUrl) {
```

Comment:
This is the same behaviour as previously, since the `urlSignal.subscribe` function would only trigger when the url changed